[CONTENT] Module 15 Project - React Tests (#1)
betrybe/sd-01-week15-react-tests: README.md

## O que deverá ser desenvolvido

Nesse projeto você escreverá testes para uma aplicação React.

## Desenvolvimento

Este repositório já contém um _template_ de uma aplicação React criado e configurado. Após clonar o projeto e instalar as dependências (mais sobre isso abaixo), você não precisará realizar nenhuma configuração adicional. Você deverá utilizar [`Jest`](https://jestjs.io/) e a biblioteca [`React Testing Library`](https://testing-library.com/) (também já instaladas e configuradas) para escrever os testes. Note que o _template_ contém uma implementação completa de todos os requisitos da Pokédex. Seu trabalho será, para cada requisito ou sub-requisito listado a seguir, escrever ao menos um teste que garanta sua corretude. Cuidado com [testes _falsos positivos_](https://talkingabouttesting.com/2015/08/04/falsos-negativos-falsos-positivos-verdadeiros-negativos-e-verdadeiros-positivos/). Falsos positivos serão desconsiderados na avaliação.

## Requisitos do projeto

A seguir estão listados todos os requisitos do projeto. Lembre-se: Cada requisito deve ser mapeado para **pelo menos** um teste. **Todos** os sub-requisitos devem também ser testados, ao menos uma vez.

> ### 1 - Ao carregar a aplicação no caminho de URL “/”, deve ser mostrado a Pokédex.

Reviewer: ```suggestion
### 1 - Ao carregar a aplicação no caminho de URL “/”, a página principal da Pokédex deve ser mostrada.
```

### 19 - Ao clicar no link "About" na barra de navegação, a aplicação deve ser redirecionada para a página de `About`, na URL "/about"

### 20 - Ao clicar no link "Favorite Pokémons" na barra de navegação, a aplicação deve ser redirecionada para a página de pokémons favoritados, na URL "/favorites"

### 21 - A página "About" deve exibir informações sobre a Pokédex

  - A página deve conter um heading `h2` com o texto `About Pokédex`;

  - A página deve conter dois parágrafos com texto sobre a Pokédex;

  - A página deve conter a seguinte imagem de uma Pokédex: `https://cdn.bulbagarden.net/upload/thumb/8/86/Gen_I_Pok%C3%A9dex.png/800px-Gen_I_Pok%C3%A9dex.png`.

### 22 - A página de pokémon favoritos deve exibir os pokémons favoritos

  - A página deve exibir todos os pokémons favoritados;

  - A página não deve exibir nenhum pokémon não favoritado.

### 23 - Entrar em uma URL desconhecida exibe a página `Not Found`

  - A página deve conter um heading `h2` com o texto `Page requested not found 😭`;

  - A página deve exibir a imagem `https://testing-library.com/`.

### 24 - A cobertura de testes deve ser 100%

>   - Para ver a cobertura de testes, execute no terminal o comando `npm run test-coverage`.

Reviewer 2: At least 44 tests? 🤔 (counting the `###` requisite and the first `-` subrequisite as one case)

Is this information relevant for the student? Like, you should develop at least 45 test cases.

And/Or, you should have a coverage above 90% 

What do you think?

Reviewer 2: forget about it. It's written right above 😂

Author: I didn't get 😂 This intention was to make the students worry about coverage. For them to succeed in this requirement, they don't need necessarily to implement all the tests.

## BÔNUS

A Pokédex é uma aplicação estática, com seus dados pré-definidos. Utilizando a [PokéAPI](https://pokeapi.co/), é possível deixá-la mais dinâmica e realista.

> Implemente os requisitos propostos as seguir e escreva testes para eles. Tente manter sempre a cobertura de testes em 100%, garantindo assim que não há código não testado. Para um desafio adicional, tente utilizar TDD - escreva os testes à medida que for implementando os requisitos.

Reviewer: ```suggestion
Implemente os requisitos propostos a seguir e escreva testes para eles. Tente manter sempre a cobertura de testes em 100%, garantindo assim que não há código ou fluxos lógicos não testados. Para um desafio adicional, tente utilizar TDD - escreva os testes à medida que for implementando os requisitos.
```

Por default, os _endpoints_ da PokéAPI retornam no máximo 20 resultados. Utilizando os parâmetros `limit` e `offset`, você pode especificar qual página deseja buscar e qual seu tamanho. Veja [aqui](https://pokeapi.co/docs/v2.html/#resource-lists-section) como utilizar estes parâmetros.

  - Adicone dois botões "Anterior" e "Próxima" que permitam navegar entre as página da lista de localizações;

  - Na primeira página, o botão "Anterior" deve ser desabilitado. Da mesma forma, ao alcançar a última página, o botão "Próximo" deve ser desabilitado.

### 28 - Adiciona uma rota para exibir uma lista de gerações

  - A URL da rota deve ser `/generations`;

  - A página deve exibir uma lista com as gerações retornadas pela PokéAPI. Você pode ler [aqui](https://pokeapi.co/docs/v2.html/#resource-lists-section) e [aqui](https://pokeapi.co/docs/v2.html/#generations) como utilizar a PokéAPI para buscar uma lista de gerações.

### 29 - Adicione na barra de navegação um link para a lista de gerações

  - O link deve conter o texto `Generations`;

  - Ao clicar no link, a página com a lista de gerações deve ser exibida.

### 30 - Adicione uma rota para exibir informações sobre uma geração

>   - A URL da rota deve ser `/generations/<id>`, onde `<id>` é o id da geração exibida.;

Reviewer 3: ```suggestion
  - A URL da rota deve ser `/generations/<id>`, onde `<id>` é o id da geração exibida;
```
